staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shufflesharding
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strings"
+)
+
+const maxHashBits = 60
+
+// ValidateParameters finds errors in the parameters for shuffle
+// sharding.  Returns a slice for which `len()` is 0 if and only if
+// there are no errors.  The entropy requirement is evaluated in a
+// fast but approximate way: bits(deckSize^handSize).
+func ValidateParameters(deckSize, handSize int) (errs []string) {
+	if handSize <= 0 {
+		errs = append(errs, "handSize is not positive")
+	}
+	if deckSize <= 0 {
+		errs = append(errs, "deckSize is not positive")
+	}
+	if len(errs) > 0 {
+		return
+	}
+	if handSize > deckSize {
+		return []string{"handSize is greater than deckSize"}
+	}
+	if math.Log2(float64(deckSize))*float64(handSize) > maxHashBits {
+		return []string{fmt.Sprintf("more than %d bits of entropy required", maxHashBits)}
+	}
+	return
+}
+
+// ShuffleAndDeal can shuffle a hash value to handSize-quantity and non-redundant
+// indices of decks, with the pick function, we can get the optimal deck index
+// Eg. From deckSize=128, handSize=8, we can get an index array [12 14 73 18 119 51 117 26],
+// then pick function will choose the optimal index from these
```

Comment:
This comment reflects an incomplete transition from the original context of fair queueing to a general-purpose utility.  I recommend completing the transition.  The word "index" is used here, but for general purpose usage we should follow the metaphor of shuffle-sharding.  This is about a "deck" of "cards".  Thus, the comment should be more like the following.

> ShuffleAndDeal shuffles a card deck of the given size and deals a hand of the given size, using the
> given hashValue as a source of entropy.
> This function makes handSize sequential calls to pick, one for each card.
> Each card is identified by an integer in the range [0, deckSize).
> For example, for deckSize=128 and handSize=4 this function might call pick(14); pick(73); pick(119); pick(26).

Reply:
The use in fair queuing does not call the one that returns a slice.
Defining a function to return a slice is the worst way to handle memory.
Better to have the caller allocate the memory.
Better still, if you easily can, to not involve a slice at all.

Reply:
OK, if I imagine a little harder I can see how the function can work if you just want to pick the "best" one in some sense.

Less memory allocs is usually better and it is easy to let the caller preallocate, so it's fine to accept a slice as well as construct one (I hadn't seen that version yet when I made this comment). It's hard for me to call this important yet without profiling the system as a whole, I'm optimizing for readability at this stage.

(My guess, not that we're at the stage to be optimizing for this yet, is that caller providing a pre-allocated slice is fastest, followed by providing a pick() function, followed by allocating a slice.)

Reply:
I think having the fair queuing client pass a `pick` function makes for the most direct source code, so I am fine with it being that way now.

Reply:
+1 valuing readability over memory-allocs here. we can refactor the pick function to `func(cards []int)` and replacing the word "index" with "card".

Reply:
The most direct code is what's there now.  Adding a slice of ints anywhere makes the code more complicated: the slice has to be allocated, the slice has to be populated, and the consumer has to loop over the members of the slice.  The current code is simpler: the implementation of `pick` is the body of that loop, and all the other overhead does not exist.

Reply:
Why pre-allocated slice is fastest here? I think we can also use pre-allocated slice in pick function. So we should keep pick(int) here? @MikeSpreitzer @yue9944882 @lavalamp

Reply:
> Mike's comment suggestion is a large improvement.
> 
> Is there a reason for this function to be public? I am having a hard time imagining when I would call this and not the one that gives me the slice.

I think this function is the basic function, which means that others can extend this function by using different pick functions, so I think it's necessary to keep this public.